File: src/main/java/sprykerFileCreator/model/renderer/PhpClassRenderer.java

package pav.sprykerFileCreator.model.renderer;

import com.intellij.openapi.util.io.StreamUtil;
import com.intellij.openapi.util.text.StringUtil;
import pav.sprykerFileCreator.model.renderer.dto.DocBlockItem;
import pav.sprykerFileCreator.model.renderer.dto.PhpClassInterface;
import pav.sprykerFileCreator.model.renderer.dto.UseBlockItem;

import java.io.IOException;
import java.util.List;

public class PhpClassRenderer implements PhpClassRendererInterface {

    private static final String NAMESPACE = "{{ namespace }}";
    private static final String USE_BLOCK = "{{ useBlock }}\n";
    private static final String DOC_BLOCK = "{{ docBlock }}\n";
    private static final String CLASS_NAME = "{{ className }}";
    private static final String PARENT_CLASS = "{{ parentClass }}";
    private static final String INTERFACE_BLOCK = "{{ interfaceBlock }}";
    private static final String METHOD_BLOCK = "{{ methodBlock }}";
    private static final String TEMPLATE_PATH = "template/PhpClassTemplate.php";


    @Override
    public String renderPhpClass(PhpClassInterface phpClass) {
        String content = this.readFile(TEMPLATE_PATH);

        if (content != null && !content.equals("")) {
            content = content.replace(NAMESPACE, phpClass.getNamespace());
            content = content.replace(USE_BLOCK, this.renderUseBlock(phpClass));
            content = content.replace(DOC_BLOCK, this.renderDocBlock(phpClass.getDocBlockItems()));
            content = content.replace(CLASS_NAME, phpClass.getName());
            content = content.replace(PARENT_CLASS, this.renderParent(phpClass));
            content = content.replace(INTERFACE_BLOCK, this.generateInterfaceBlock(phpClass));
            content = content.replace(METHOD_BLOCK, this.generateMethodBlock(phpClass));
        }

        return content;
    }

    public String renderDocBlock(List<DocBlockItem> docBlockItems) {

        if (docBlockItems.size() == 0) {
            return "";
        }

        String docBlock = "/**\n";

        for (DocBlockItem docBlockItem: docBlockItems) {
            String[] docBlockElements = {
                    docBlockItem.getTag(),
                    docBlockItem.getReturnType(),
                    docBlockItem.getValue(),
            };
            docBlock += " * @" + StringUtil.join(docBlockElements, " ") + "\n";
        }

        docBlock += " */\n";

        return docBlock;
    }


    private String renderUseBlock(PhpClassInterface phpClass) {

        List<UseBlockItem> useBlockItems = phpClass.getUseBlockItems();

        if (useBlockItems.size() == 0) {
            return "";
        }

        String useBlock = "";

        for (UseBlockItem useItem : phpClass.getUseBlockItems()) {
            useBlock += "use " + useItem.getFullQualifiedClassName();

            if (useItem.getAlias() != null) {
                useBlock += " as " + useItem.getAlias();
            }

            useBlock += ";\n";
        }

        return useBlock + "\n";
    }

    private String renderParent(PhpClassInterface phpClass) {
        PhpClassInterface parentClass = phpClass.getParentClass();

        if (parentClass == null) {
            return "";
        }

        String parentClassName = parentClass.getName();

        if (!phpClass.getParentAlias().equals("")) {
            parentClassName = phpClass.getParentAlias();
        }

        return " extends " + parentClassName;
    }

    private String generateMethodBlock(PhpClassInterface phpClass) {
        return "";
    }

    private String generateInterfaceBlock(PhpClassInterface phpClass) {
        return "";
    }

    /**
     *
     * @param path String
     * @return String
     */
    private String readFile(String path)
    {
        String content = "";

        try {
            content = StreamUtil.readText(
                    PhpClassRenderer.class.getResourceAsStream(path),
                    "UTF-8"
            );
        } catch (IOException e) {
            e.printStackTrace();
        }

        return content;
    }

}

File: src/main/kotlin/com/github/tobi812/sprykerplugin/services/renderer/PhpClassRenderer.kt
package com.github.tobi812.sprykerplugin.services.renderer

import com.github.tobi812.sprykerplugin.services.renderer.dto.DocBlockItem
import com.github.tobi812.sprykerplugin.services.renderer.dto.PhpClassInterface
import com.github.tobi812.sprykerplugin.services.renderer.dto.UseBlockItem
import com.intellij.openapi.util.text.StringUtil
import com.jetbrains.php.lang.psi.elements.Method
import com.jetbrains.php.lang.psi.elements.PhpClass
import java.io.IOException

class PhpClassRenderer : PhpClassRendererInterface {
    companion object {
        private const val NAMESPACE = "{{ namespace }}"
        private const val USE_BLOCK = "{{ useBlock }}\n"
        private const val DOC_BLOCK = "{{ docBlock }}\n"
        private const val CLASS_NAME = "{{ className }}"
        private const val PARENT_CLASS = "{{ parentClass }}"
        private const val INTERFACE_BLOCK = "{{ interfaceBlock }}"
        private const val METHOD_BLOCK = "{{ methodBlock }}"
        private const val TEMPLATE_PATH = "/template/PhpClassTemplate.php.template"
    }

    override fun renderPhpClass(phpClass: PhpClassInterface): String {
        var content = this.readFile(TEMPLATE_PATH)

        if (content != "") {
            content = content.replace(NAMESPACE, phpClass.namespace)
            content = content.replace(USE_BLOCK, this.renderUseBlock(phpClass))
            content = content.replace(DOC_BLOCK, this.renderDocBlock(phpClass.docBlockItems))
            content = content.replace(CLASS_NAME, phpClass.name)
            content = content.replace(PARENT_CLASS, renderParent(phpClass))
            content = content.replace(INTERFACE_BLOCK, generateInterfaceBlock(phpClass))
            content = content.replace(METHOD_BLOCK, generateMethodBlock(phpClass))
        }

        return content
    }

    override fun renderDocBlock(docBlockItems: List<DocBlockItem>): String {
        if (docBlockItems.isEmpty()) {
            return ""
        }

        var docBlock = "/**\n"
        for (docBlockItem in docBlockItems) {
            val docBlockElements = ArrayList<String>()
            docBlockElements.add(docBlockItem.tag)
            docBlockElements.add(docBlockItem.returnType)

            if (docBlockItem.value != null) {
                docBlockElements.add(docBlockItem.value)
            }

            docBlock += " * @${StringUtil.join(docBlockElements, " ")}\n"
        }

        docBlock += " */\n"

        return docBlock
    }

    override fun renderFactoryMethod(phpClass: PhpClass): String {
        var method: String = "public function create"
        method += phpClass.name
        method += "()\n{\n"
        method += "    return new ${phpClass.name}("

        val constructor = phpClass.constructor
        if (constructor is Method) {
            method += this.renderConstructorCall(constructor, phpClass)
        }

        method += "\n     );\n}"

        return method
    }

    private fun renderConstructorCall(
        constructor: Method,
        phpClass: PhpClass
    ): String {
        val methodCalls = ArrayList<String>()
        val parameters = constructor.parameters
        for (parameter in parameters) {
            if (parameter.declaredType.isEmpty) {
                methodCalls.add(this.createMethodCallString(parameter.name.capitalize()))

                continue
            }

            val classNamespace = this.extractModuleNamespace(phpClass.namespaceName)
            val parameterNamespace = this.extractModuleNamespace(parameter.declaredType.toString())

            val parameterType = parameter.localType.toString().split("\\").last()

            var methodPrefix = "create"
            if (classNamespace != parameterNamespace) {
                methodPrefix = "get"
            }

            methodCalls.add(this.createMethodCallString(parameterType, methodPrefix))
        }

        return StringUtil.join(methodCalls, ",")
    }

    private fun renderUseBlock(phpClass: PhpClassInterface): String {
        val useBlockItems: List<UseBlockItem> = phpClass.useBlockItems
        if (useBlockItems.isEmpty()) {
            return ""
        }

        var useBlock = ""
        for (useItem in phpClass.useBlockItems) {
            useBlock += "use " + useItem.fullQualifiedClassName
            if (useItem.alias != null) {
                useBlock += " as " + useItem.alias
            }
            useBlock += ";\n"
        }

        return """
            $useBlock
            
            """.trimIndent()
    }

    private fun renderParent(phpClass: PhpClassInterface): String {
        val parentClass: PhpClassInterface = phpClass.parentClass ?: return ""
        var parentClassName: String = parentClass.name

        val parentAlias: String = phpClass.parentAlias as String
        if (parentAlias.isNotBlank()) parentClassName = parentAlias

        return " extends $parentClassName"
    }

    private fun generateMethodBlock(phpClass: PhpClassInterface): String {
        return ""
    }

    private fun generateInterfaceBlock(phpClass: PhpClassInterface): String {
        return ""
    }

    private fun readFile(path: String): String {
        try {
            val resource = javaClass.getResource(path)

            if (resource != null) {
                return resource.readText()
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return ""
    }

    private fun extractModuleNamespace(namespace: String): String {
        val preparedNamespace = namespace.removePrefix("\\")
        val namespaceSegments = preparedNamespace.split("\\")
        val appName = namespaceSegments[1]
        var occurence = 3

        if (appName == "Zed") {
            occurence = 4
        }

        val index = this.indexOf(preparedNamespace, "\\", occurence)

        return preparedNamespace.substring(0, index).removeSuffix("\\")
    }

    private fun indexOf(string: String, subString: String, occurence: Int): Int {
        var i = 0
        var match = 0
        while (i < occurence) {
            match = string.indexOf(subString, match) + 1
            i++
        }

        return match
    }

    private fun createMethodCallString(className: String, methodPrefix: String = "create"): String {
        val preparedClassName = className
            .removeSuffix("Interface")
            .removePrefix("Abstract")

        return "\n\$this->${methodPrefix}${preparedClassName}()"
    }
}
